File: sdks/python/tests/test_sdk.py
import time
import keelson

from keelson.payloads.Primitives_pb2 import TimestampedFloat


def test_construct_pub_sub_key():
    assert (
        keelson.construct_pubsub_key(
            base_path="base_path",
            entity_id="entity_id",
            subject="subject",
            source_id="source_id",
        )
        == "base_path/@v0/entity_id/pubsub/subject/source_id"
    )


def test_construct_rpc_key():
    assert (
        keelson.construct_rpc_key(
            base_path="base_path",
            entity_id="entity_id",
            procedure="procedure",
            responder_id="responder_id",
        )
        == "base_path/@v0/entity_id/@rpc/procedure/responder_id"
    )


def test_parse_pub_sub_key():
    assert keelson.parse_pubsub_key(
        "base_path/@v0/entity_id/pubsub/subject/source_id/sub_id"
    ) == dict(
        base_path="base_path",
        entity_id="entity_id",
        subject="subject",
        source_id="source_id/sub_id",
        target_id=None,
    )


def test_parse_rpc_key():
    assert keelson.parse_rpc_key(
        "base_path/@v0/entity_id/@rpc/procedure/responder_id"
    ) == dict(
        base_path="base_path",
        entity_id="entity_id",
        procedure="procedure",
        responder_id="responder_id",
    )


def test_get_subject_from_pub_sub_key():
    assert (
        keelson.get_subject_from_pubsub_key(
            "base_path/@v0/entity_id/pubsub/subject/source_id"
        )
        == "subject"
    )


def test_enclose_uncover():
    test = b"test"
    message = keelson.enclose(payload=test)
    received_at, enclosed_at, payload = keelson.uncover(message)

    assert test == payload
    assert received_at >= enclosed_at


def test_enclose_uncover_actual_payload():
    data = TimestampedFloat()
    data.timestamp.FromNanoseconds(time.time_ns())
    data.value = 3.14
    message = keelson.enclose(data.SerializeToString())
    received_at, enclosed_at, payload = keelson.uncover(message)
    content = TimestampedFloat.FromString(payload)

    assert data.value == content.value
    assert data.timestamp == content.timestamp
    assert received_at >= enclosed_at


def test_get_protobuf_file_descriptor_set_from_type_name():
    file_descriptor_set = keelson.get_protobuf_file_descriptor_set_from_type_name(
        "keelson.TimestampedString"
    )
    assert file_descriptor_set


def test_decode_protobuf_using_generated_message_classes():
    data = TimestampedFloat()
    data.timestamp.FromNanoseconds(time.time_ns())
    data.value = 3.14

    payload = data.SerializeToString()

    decoded = keelson.decode_protobuf_payload_from_type_name(
        payload, "keelson.TimestampedFloat"
    )

    assert data.value == decoded.value
    assert (
        data.timestamp.ToNanoseconds()
        == decoded.timestamp.ToNanoseconds()
        # These are different class definitions and will fail a direct comparison...
    )


def test_ensure_all_well_known_tags():
    for subject, proto_type_name in keelson._SUBJECTS.items():
        assert subject == str(subject).lower()

        assert keelson.get_protobuf_file_descriptor_set_from_type_name(proto_type_name)


def test_is_subject_well_known():
    assert keelson.is_subject_well_known("lever_position_pct") is True
    assert keelson.is_subject_well_known("random_mumbo_jumbo") is False


def test_get_subject_schema():
    assert (
        keelson.get_subject_schema("lever_position_pct") == "keelson.TimestampedFloat"
    )


def test_construct_liveliness_key():
    assert (
        keelson.construct_liveliness_key(
            base_path="keelson",
            entity_id="landkrabban",
            source_id="gnss/0",
        )
        == "keelson/@v0/landkrabban/pubsub/*/gnss/0"
    )


def test_parse_liveliness_key():
    key = keelson.construct_liveliness_key(
        base_path="keelson",
        entity_id="landkrabban",
        source_id="gnss/0",
    )
    parsed = keelson.parse_liveliness_key(key)
    assert parsed == dict(
        base_path="keelson",
        entity_id="landkrabban",
        source_id="gnss/0",
    )


def test_parse_liveliness_key_with_slashed_source():
    parsed = keelson.parse_liveliness_key("keelson/@v0/landkrabban/pubsub/*/gnss/0")
    assert parsed["source_id"] == "gnss/0"


def test_parse_liveliness_key_invalid():
    import pytest

    with pytest.raises(ValueError):
        keelson.parse_liveliness_key("keelson/@v0/entity/pubsub/some_subject/source")


def test_subpackages_importability():
    pass


def test_interfaces_importability():
    pass


# Tests for @target extension


def test_construct_pubsub_key_with_target_id():
    """Test constructing a pubsub key with target_id."""
    assert (
        keelson.construct_pubsub_key(
            base_path="keelson",
            entity_id="shore_station",
            subject="heading_true_north_deg",
            source_id="ais",
            target_id="mmsi_245060000",
        )
        == "keelson/@v0/shore_station/pubsub/heading_true_north_deg/ais/@target/mmsi_245060000"
    )


def test_construct_pubsub_key_with_target_id_none():
    """Test that target_id=None produces same result as omitting it."""
    key_without = keelson.construct_pubsub_key(
        base_path="keelson",
        entity_id="entity",
        subject="subject",
        source_id="source",
    )
    key_with_none = keelson.construct_pubsub_key(
        base_path="keelson",
        entity_id="entity",
        subject="subject",
        source_id="source",
        target_id=None,
    )
    assert key_without == key_with_none
    assert "@target" not in key_without


def test_construct_pubsub_key_with_slashed_source_and_target():
    """Test constructing a pubsub key with nested source_id and target_id."""
    assert (
        keelson.construct_pubsub_key(
            base_path="keelson",
            entity_id="vessel",
            subject="location_fix",
            source_id="ais/receiver/0",
            target_id="mmsi_123456789",
        )
        == "keelson/@v0/vessel/pubsub/location_fix/ais/receiver/0/@target/mmsi_123456789"
    )


def test_parse_pubsub_key_with_target_id():
    """Test parsing a pubsub key with @target extension."""
    parsed = keelson.parse_pubsub_key(
        "keelson/@v0/shore_station/pubsub/heading_true_north_deg/ais/@target/mmsi_245060000"
    )
    assert parsed == dict(
        base_path="keelson",
        entity_id="shore_station",
        subject="heading_true_north_deg",
        source_id="ais",
        target_id="mmsi_245060000",
    )


def test_parse_pubsub_key_without_target_id():
    """Test parsing a pubsub key without @target extension returns target_id=None."""
    parsed = keelson.parse_pubsub_key(
        "keelson/@v0/entity/pubsub/subject/source"
    )
    assert parsed == dict(
        base_path="keelson",
        entity_id="entity",
        subject="subject",
        source_id="source",
        target_id=None,
    )


def test_parse_pubsub_key_with_slashed_source_and_target():
    """Test parsing a pubsub key with nested source_id and target_id."""
    parsed = keelson.parse_pubsub_key(
        "keelson/@v0/vessel/pubsub/location_fix/ais/receiver/0/@target/mmsi_123456789"
    )
    assert parsed == dict(
        base_path="keelson",
        entity_id="vessel",
        subject="location_fix",
        source_id="ais/receiver/0",
        target_id="mmsi_123456789",
    )


def test_pubsub_key_roundtrip_with_target():
    """Test that construct -> parse -> construct produces identical key."""
    original_key = keelson.construct_pubsub_key(
        base_path="keelson",
        entity_id="shore_station",
        subject="speed_over_ground_knots",
        source_id="ais/receiver",
        target_id="mmsi_987654321",
    )
    parsed = keelson.parse_pubsub_key(original_key)
    reconstructed_key = keelson.construct_pubsub_key(
        base_path=parsed["base_path"],
        entity_id=parsed["entity_id"],
        subject=parsed["subject"],
        source_id=parsed["source_id"],
        target_id=parsed["target_id"],
    )
    assert original_key == reconstructed_key


def test_pubsub_key_roundtrip_without_target():
    """Test that construct -> parse -> construct produces identical key (no target)."""
    original_key = keelson.construct_pubsub_key(
        base_path="keelson",
        entity_id="landkrabban",
        subject="location_fix",
        source_id="gnss/0",
    )
    parsed = keelson.parse_pubsub_key(original_key)
    reconstructed_key = keelson.construct_pubsub_key(
        base_path=parsed["base_path"],
        entity_id=parsed["entity_id"],
        subject=parsed["subject"],
        source_id=parsed["source_id"],
        target_id=parsed["target_id"],
    )
    assert original_key == reconstructed_key


def test_get_subject_from_pubsub_key_with_target():
    """Test get_subject_from_pubsub_key works with @target extension."""
    assert (
        keelson.get_subject_from_pubsub_key(
            "keelson/@v0/shore_station/pubsub/heading_true_north_deg/ais/@target/mmsi_245060000"
        )
        == "heading_true_north_deg"
    )

File: sdks/rust/tests/test_sdk.rs
use keelson::*;
use keelson::payloads::TimestampedFloat;
use prost::Message;
use prost_types::Timestamp;
use std::time::{SystemTime, UNIX_EPOCH};

#[test]
fn test_construct_pubsub_key() {
    let key = construct_pubsub_key("base_path", "entity_id", "subject", "source_id");
    assert_eq!(key, "base_path/@v0/entity_id/pubsub/subject/source_id");
}

#[test]
fn test_construct_rpc_key() {
    let key = construct_rpc_key("base_path", "entity_id", "procedure", "responder_id");
    assert_eq!(key, "base_path/@v0/entity_id/@rpc/procedure/responder_id");
}

#[test]
fn test_parse_pubsub_key() {
    let key = "base_path/@v0/entity_id/pubsub/subject/source_id/sub_id";
    let parsed = parse_pubsub_key(key).unwrap();
    assert_eq!(parsed.0, "base_path");
    assert_eq!(parsed.1, "entity_id");
    assert_eq!(parsed.2, "subject");
    assert_eq!(parsed.3, "source_id/sub_id");
}

#[test]
fn test_parse_rpc_key() {
    let key = "base_path/@v0/entity_id/@rpc/procedure/responder_id";
    let parsed = parse_rpc_key(key).unwrap();
    assert_eq!(parsed.0, "base_path");
    assert_eq!(parsed.1, "entity_id");
    assert_eq!(parsed.2, "procedure");
    assert_eq!(parsed.3, "responder_id");
}

#[test]
fn test_get_subject_from_pubsub_key() {
    let key = "base_path/@v0/entity_id/pubsub/subject/source_id";
    let subject = get_subject_from_pubsub_key(key).unwrap();
    assert_eq!(subject, "subject");
}

#[test]
fn test_enclose_uncover() {
    let test_data = b"test".to_vec();
    let envelope = enclose(test_data.clone(), None);
    let (received_at, enclosed_at, payload) = uncover(&envelope).unwrap();

    assert_eq!(test_data, payload);
    assert!(received_at.seconds >= enclosed_at.as_ref().unwrap().seconds);
}

#[test]
fn test_enclose_uncover_actual_payload() {
    let mut data = TimestampedFloat::default();
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
    data.timestamp = Some(Timestamp {
        seconds: now.as_secs() as i64,
        nanos: now.subsec_nanos() as i32,
    });
    data.value = 3.14;

    let envelope = enclose(data.encode_to_vec(), None);
    let (received_at, enclosed_at, payload) = uncover(&envelope).unwrap();
    let content = TimestampedFloat::decode(&payload[..]).unwrap();

    assert_eq!(data.value, content.value);
    assert_eq!(data.timestamp, content.timestamp);
    assert!(received_at.seconds >= enclosed_at.unwrap().seconds);
}

#[test]
fn test_is_subject_well_known() {
    assert!(is_subject_well_known("lever_position_pct"));
    assert!(!is_subject_well_known("random_mumbo_jumbo"));
}

#[test]
fn test_get_subject_schema() {
    assert_eq!(get_subject_schema("lever_position_pct"), Some("keelson.TimestampedFloat"));
    assert_eq!(get_subject_schema("raw"), Some("keelson.TimestampedBytes"));
    assert_eq!(get_subject_schema("not_a_subject"), None);
}

#[test]
fn test_decode_payload_from_type_name() {
    let mut data = TimestampedFloat::default();
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
    data.timestamp = Some(Timestamp {
        seconds: now.as_secs() as i64,
        nanos: now.subsec_nanos() as i32,
    });
    data.value = 3.14;

    let payload = data.encode_to_vec();
    let decoded: TimestampedFloat = decode_payload_from_type_name(&payload).unwrap();

    assert_eq!(data.value, decoded.value);
    assert_eq!(data.timestamp, decoded.timestamp);
}

#[test]
fn test_encode_payload_from_type_name() {
    let mut data = TimestampedFloat::default();
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
    data.timestamp = Some(Timestamp {
        seconds: now.as_secs() as i64,
        nanos: now.subsec_nanos() as i32,
    });
    data.value = 3.14;

    let encoded = encode_payload_from_type_name(&data);
    let decoded = TimestampedFloat::decode(&encoded[..]).unwrap();

    assert_eq!(data.value, decoded.value);
    assert_eq!(data.timestamp, decoded.timestamp);
}

#[test]
fn test_enclose_from_type_name() {
    let mut data = TimestampedFloat::default();
    let now = SystemTime::now().duration_since(UNIX_EPOCH).unwrap();
    data.timestamp = Some(Timestamp {
        seconds: now.as_secs() as i64,
        nanos: now.subsec_nanos() as i32,
    });
    data.value = 3.14;

    let envelope = enclose_from_type_name(&data, None);
    let (received_at, enclosed_at, payload) = uncover(&envelope).unwrap();
    let content = TimestampedFloat::decode(&payload[..]).unwrap();

    assert_eq!(data.value, content.value);
    assert_eq!(data.timestamp, content.timestamp);
    assert!(received_at.seconds >= enclosed_at.unwrap().seconds);
}

#[test]
fn test_ensure_all_well_known_subjects() {
    // Test that all subjects are lowercase and have valid schemas
    let test_subjects = ["lever_position_pct", "raw", "heading_true_north_deg"];
    
    for subject in &test_subjects {
        assert_eq!(*subject, subject.to_lowercase());
        assert!(is_subject_well_known(subject), "Subject '{}' should be well-known", subject);
        assert!(get_subject_schema(subject).is_some(), "Subject '{}' should have a schema", subject);
    }
}

// Test codec functionality
#[test]
fn test_codec_text_roundtrip() {
    use keelson::codec::{enclose_from_text, uncover_to_text};
    
    let key = "test/@v0/entity/pubsub/raw/source";
    let original_text = "Hello, Keelson!";
    
    let encoded = enclose_from_text(key, original_text).unwrap();
    let decoded = uncover_to_text(key, &encoded).unwrap();
    
    assert_eq!(original_text, decoded);
}

#[test]
fn test_codec_base64_roundtrip() {
    use keelson::codec::{enclose_from_base64, uncover_to_base64};
    
    let key = "test/@v0/entity/pubsub/raw/source";
    let original_data = "Hello, World!";
    use base64::{Engine as _, engine::general_purpose};
    let base64_data = general_purpose::STANDARD.encode(original_data);
    
    let encoded = enclose_from_base64(key, &base64_data).unwrap();
    let decoded = uncover_to_base64(key, &encoded).unwrap();
    
    assert_eq!(base64_data, decoded);
}
